File: advent_of_code/utils/read_file_lines.go
package utils

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// ReadFileLines reads a text file and returns a slice of its non-empty lines.
func ReadFileLines(filePath string) ([]string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("error opening file: %v", err)
	}
	defer file.Close()

	// Read the file line by line and filter non-empty lines.
	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" {
			lines = append(lines, line)
		}
	}

	// Check for any scanning errors.
	if err := scanner.Err(); err != nil {
		fmt.Println("error scanning file:", err)
		return nil, fmt.Errorf("error reading file: %v", err)
	}

	return lines, nil
}

File: advent_of_code/utils/read_file_lines.py
def read_file_lines(file_path: str) -> list[str]:
    """Reads a text file and returns a list of its non-empty lines.
    Args:
        file_path: The relative path to the text file.
    Returns:
        A list of non-empty lines in the file, with any trailing newline characters removed.
    Raises:
        IOError: If there is an error opening or reading the file.
    """
    try:
        with open(file_path, "r") as file:
            lines = file.readlines()
            return [line.rstrip("\n") for line in lines if line.rstrip("\n")]
    except IOError as e:
        raise IOError(f"Error reading file: {e}") from e
    except Exception as e:
        raise IOError(f"An unexpected error occurred: {e}") from e
